src/ModelContextProtocol/Protocol/Transport/StreamableHttpClientSessionTransport.cs

```diff
@@ -0,0 +1,267 @@
+    {
+        // Send a GET request to handle any unsolicited messages not sent over a POST response.
+        using var request = new HttpRequestMessage(HttpMethod.Get, _options.Endpoint);
+        request.Headers.Accept.Add(TextEventStreamMediaType);
+        CopyAdditionalHeaders(request.Headers, _options.AdditionalHeaders, _mcpSessionId);
+
+        using var response = await _httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, _connectionCts.Token).ConfigureAwait(false);
+
+        if (!response.IsSuccessStatusCode)
+        {
+            // Server support for the GET request is optional. If it fails, we don't care. It just means we won't receive unsolicited messages.
+            return;
+        }
+
+        response.EnsureSuccessStatusCode();
+        var responseStream = await response.Content.ReadAsStreamAsync(_connectionCts.Token).ConfigureAwait(false);
+        await ProcessSseResponseAsync(responseStream, relatedRpcRequest: null, _connectionCts.Token).ConfigureAwait(false);
+    }
+
+    private async Task<JsonRpcMessageWithId?> ProcessSseResponseAsync(Stream responseStream, JsonRpcRequest? relatedRpcRequest, CancellationToken cancellationToken)
+    {
+        await foreach (SseItem<string> sseEvent in SseParser.Create(responseStream).EnumerateAsync(cancellationToken).ConfigureAwait(false))
+        {
+            if (sseEvent.EventType != "message")
+            {
+                continue;
+            }
+
+            var message = await ProcessMessageAsync(sseEvent.Data, cancellationToken).ConfigureAwait(false);
+
+            // The server SHOULD end the response here anyway, but we won't leave it to chance. This transport makes
+            // a GET request for any notifications that might need to be sent after the completion of each POST.
+            if (message is JsonRpcMessageWithId messageWithId && relatedRpcRequest?.Id == messageWithId.Id)
+            {
+                return messageWithId;
+            }
+        }
+
+        return null;
+    }
+
+    private async Task<JsonRpcMessageWithId?> ProcessJsonRpcBatch(string arrayData, JsonRpcRequest? relatedRpcRequest, CancellationToken cancellationToken)
+    {
+        try
+        {
+            var batch = JsonSerializer.Deserialize(arrayData, McpJsonUtilities.JsonContext.Default.JsonRpcMessageArray);
+            if (batch is null)
+            {
+                throw new Exception("Unreachable! We already checked arrayData started with '['");
+            }
+
+            JsonRpcMessageWithId? relatedRpcResponse = null;
+
+            foreach (var message in batch!)
+            {
+                await WriteMessageAsync(message, cancellationToken).ConfigureAwait(false);
+                if (message is JsonRpcMessageWithId messageWithId && relatedRpcRequest?.Id == messageWithId.Id)
+                {
+                    relatedRpcResponse = messageWithId;
+                }
+            }
+
+            return relatedRpcResponse;
```

Comment:
Only the last is needed by the caller?

Reply:
This isn't necessarily the last batched message, just the last message with the Id we're looking for. And since this is only used to raise an error when we don't get a response at all, I think that's fine. If there are multiple responses with the same ID, McpSession will emit an log indicating it got a message with an unknown ID.